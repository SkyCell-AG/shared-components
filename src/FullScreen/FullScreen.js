import React, {
    useRef,
    useCallback,
    useState,
    useEffect,
} from 'react'
import noop from 'lodash/noop'
import PropTypes from 'prop-types'
import clsx from 'clsx'
import FullscreenIcon from '@mui/icons-material/Fullscreen'
import FullscreenExitIcon from '@mui/icons-material/FullscreenExit'

import useStyles from './FullScreen.styles'

const propTypes = {
    className: PropTypes.string,
    children: PropTypes.element.isRequired,
    onFullScreen: PropTypes.func,
    setTemperatureChartFullscreen: PropTypes.func,
}

const defaultProps = {
    className: '',
    onFullScreen: noop,
    setTemperatureChartFullscreen: noop,
}

const FullScreen = ({
    className,
    children,
    onFullScreen,
    setTemperatureChartFullscreen,
}) => {
    const classes = useStyles()
    const wrapper = useRef(null)
    const [
        fullScreen,
        setFullScreen,
    ] = useState(false)

    const requestFullScreen = useCallback(() => {
        let prevented = false
        const event = {
            preventDefault: () => {
                prevented = true
            },
        }

        onFullScreen(event)
        if (prevented) {
            return
        }

        if (fullScreen) {
            document.exitFullscreen()
        } else {
            wrapper.current.requestFullscreen({
                navigationUI: 'show',
            }).then({})
        }
    }, [
        fullScreen,
        onFullScreen,
    ])

    const onFullScreenChange = useCallback(() => {
        setFullScreen(document.fullscreenElement)
        setTemperatureChartFullscreen(document.fullscreenElement)
        if (document.fullscreenElement) {
            document.body.classList.add(classes.fullscreen)
            const chartContainer = document.querySelector(`.${classes.children}`)

            chartContainer.style.display = 'flex'
            chartContainer.style.alignItems = 'center'
            chartContainer.style.height = '100%'
        } else {
            document.body.classList.remove(classes.fullscreen)

            const chartContainer = document.querySelector(`.${classes.children}`)

            chartContainer.style.display = ''
            chartContainer.style.alignItems = ''
            chartContainer.style.height = ''
        }
    }, [
        classes.children,
        classes.fullscreen,
        setTemperatureChartFullscreen,
    ])

    useEffect(() => {
        const currentWrapper = wrapper.current

        currentWrapper.addEventListener('fullscreenchange', onFullScreenChange)
        return () => {
            currentWrapper.removeEventListener('fullscreenchange', onFullScreenChange)
        }
    }, [onFullScreenChange])

    return (
        <div
            className={clsx(
                className,
                classes.wrapper,
            )}
            ref={wrapper}
        >
            <div
                className={fullScreen
                    ? classes.fullScreenButtonClose
                    : classes.fullScreenButtonOpen}
                onMouseDown={requestFullScreen}
            >
                {fullScreen ? (
                    <FullscreenExitIcon />
                ) : (
                    <FullscreenIcon />
                )}
            </div>
            <div
                className={classes.children}
            >
                {children}
            </div>
        </div>
    )
}

FullScreen.propTypes = propTypes
FullScreen.defaultProps = defaultProps

export default FullScreen
