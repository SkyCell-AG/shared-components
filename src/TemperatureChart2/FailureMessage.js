import React from 'react'
import {
    Grid,
} from '@material-ui/core'
import clsx from 'clsx'

import notFound from './not-found.png'
import useStyles from './TemperatureChart2.style'

const FailureMessage = () => {
    const classes = useStyles()

    return (
        <div className={clsx(
            classes.overlay,
            classes.noData,
        )}
        >
            <Grid
                container
                spacing={3}
            >
                <Grid
                    item
                    md={6}
                >
                    <img
                        src={notFound}
                        alt="Not Found"
                    />
                </Grid>
                <Grid
                    item
                    xs={8}
                    md={6}
                    className={classes.msg}
                >
                    <div className={classes.middle}>
                        Ouups, we&apos;re sorry -
                        it seems we don&apos;t have
                        temperature data for the selected period!
                    </div>
                </Grid>
            </Grid>
        </div>
    )
}

export default FailureMessage
