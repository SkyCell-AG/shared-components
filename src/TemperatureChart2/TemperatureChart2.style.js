import {
    makeStyles,
} from '@mui/styles'
import {
    useTheme,
} from '@mui/material/styles'

const useStyles = makeStyles(() => {
    const theme = useTheme()

    return {
        wrapper: {
            position: 'relative',
            display: 'flex',
            flexDirection: 'column',
            alignItems: 'center',
            justifyContent: 'center',
            height: '100%',
            width: '100%',
        },
        chart: {
            width: '100%',
            height: '575px',
        },
        chartPrinting: {
            width: 1020,
            height: 550,
        },
        fullscreen: {
            position: 'absolute',
            height: '85%',
            width: '100%',
        },
        overlay: {
            flexGrow: 1,
            display: 'flex',
            zIndex: '100 !important',
            width: '90%',
            height: '75%',
            backgroundColor: theme.palette.common.white,
            alignItems: 'center',
            justifyContent: 'center',
        },
        noData: {
            backgroundColor: theme.palette.common.white,
        },
        msg: {
            display: 'flex',
            alignItems: 'center',
        },
        middle: {
            margin: 0,
        },
    }
})

export default useStyles
