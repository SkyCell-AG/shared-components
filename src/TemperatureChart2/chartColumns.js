const chartColumns = [[
    'date',
    'Date',
]]

export default chartColumns
