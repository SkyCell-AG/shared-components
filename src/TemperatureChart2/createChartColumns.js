const column = [
    [{
        type: 'string', role: 'annotation',
    }],
    [{
        type: 'string', role: 'annotationText',
    }],
]

const createChartColumns = (sensorLabels) => {
    return sensorLabels ? sensorLabels.reduce((arr, label) => {
        arr.push(
            [
                'number',
                `${label} [CÂ°]`,
            ],
            ...column,
        )

        return arr
    }, []) : []
}

export default createChartColumns
