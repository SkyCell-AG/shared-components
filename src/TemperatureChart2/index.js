export {
    default,
} from './TemperatureChart2'
