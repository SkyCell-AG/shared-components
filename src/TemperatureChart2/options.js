const options = {
    chartArea: {
        left: 30,
        top: 20,
        width: '95%',
        height: '87%',
    },
    axisTitlesPosition: 'none',
    vAxis: {
        title: 'Temperature',
        gridlines: {
            count: 0,
        },
    },
    curveType: 'function',
    series: [
        {
            targetAxisIndex: 0,
            lineWidth: 3,
            color: '#CF3B8A',
            pointSize: 6,
        },
        {
            targetAxisIndex: 0,
            lineWidth: 3,
            color: '#61C6E9',
            pointSize: 6,
        },
        {
            targetAxisIndex: 0,
            lineWidth: 3,
            color: '#edae49',
            lineDashStyle: [
                4,
                4,
            ],
        },
        {
            targetAxisIndex: 0,
            lineWidth: 3,
            color: '#edae49',
            lineDashStyle: [
                4,
                4,
            ],
        },
    ],
}

export default options
