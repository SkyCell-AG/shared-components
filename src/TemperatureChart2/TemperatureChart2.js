import React, {
    useMemo,
} from 'react'
import PropTypes from 'prop-types'
import noop from 'lodash/noop'

import LineChart from 'LineChart'
import FullScreen from 'FullScreen'
import strToDate from 'utils/strToDate'

import createChartColumns from './createChartColumns'
import chartColumns from './chartColumns'
import options from './options'
import addNotationValues from './addNotationValues'

import useStyles from './TemperatureChart2.style'

const propTypes = {
    sensorData: PropTypes.arrayOf(PropTypes.shape({
        d: PropTypes.string.isRequired,
        t: PropTypes.string.isRequired,
    })).isRequired,
    sensorLabels: PropTypes.arrayOf(PropTypes.string).isRequired,
    onError: PropTypes.func,
    isChartPrinting: PropTypes.bool,
    onFullScreen: PropTypes.func,
    userOptions: PropTypes.any, // eslint-disable-line
    temperatureChartFullscreen: PropTypes.bool,
    setTemperatureChartFullscreen: PropTypes.func,
}

const defaultProps = {
    onError: noop,
    onFullScreen: noop,
    isChartPrinting: false,
    userOptions: options,
    temperatureChartFullscreen: false,
    setTemperatureChartFullscreen: noop,
}

const TemperatureChart2 = (props) => {
    const {
        sensorData,
        sensorLabels,
        isChartPrinting,
        onFullScreen,
        userOptions,
        temperatureChartFullscreen,
        setTemperatureChartFullscreen,
        onError,
    } = props

    const classes = useStyles()

    const chartArea = useMemo(() => {
        return {
            left: 20,
            top: 10,
            width: '100%',
            height: '87%',
        }
    }, [])

    const columns = useMemo(() => {
        return [
            ...chartColumns,
            ...createChartColumns(sensorLabels),
        ]
    }, [sensorLabels])

    const chartPrinting = useMemo(() => {
        return isChartPrinting
            ? classes.chartPrinting
            : classes.chart
    }, [
        classes.chart,
        classes.chartPrinting,
        isChartPrinting,
    ])

    const chartOptions = useMemo(() => {
        return isChartPrinting ? {
            ...userOptions,
            chartArea,
        } : userOptions
    }, [
        chartArea,
        isChartPrinting,
        userOptions,
    ])

    const value = useMemo(() => {
        return sensorData.map(({
            d,
            t,
        }) => {
            return [
                strToDate(t),
                ...addNotationValues(d),
            ]
        })
    }, [sensorData])

    return (
        <FullScreen
            {...props}
            className={classes.wrapper}
            onFullScreen={onFullScreen}
            setTemperatureChartFullscreen={setTemperatureChartFullscreen}
        >
            <LineChart
                value={value}
                columns={columns}
                userOptions={chartOptions}
                onError={onError}
                className={temperatureChartFullscreen
                    ? classes.fullscreen
                    : chartPrinting}
            />
        </FullScreen>
    )
}

TemperatureChart2.propTypes = propTypes
TemperatureChart2.defaultProps = defaultProps

export default TemperatureChart2
