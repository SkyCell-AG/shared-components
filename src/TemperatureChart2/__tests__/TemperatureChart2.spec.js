import React from 'react'
import {
    shallow,
} from 'enzyme'

import FullScreen from 'FullScreen'
import LineChart from 'LineChart'
import TemperatureChart2 from '../TemperatureChart2'

it('TemperatureChart success', () => {
    const wrapper = shallow(
        <TemperatureChart2
            sensorData={[
                {
                    d: [
                        18.1,
                        5.3,
                    ],
                    t: '29.03.2021 10:40',
                },
                {
                    d: [
                        18.1,
                        5.3,
                    ],
                    t: '29.03.2021 10:50',
                },
                {
                    d: [
                        18.2,
                        5.3,
                    ],
                    t: '29.03.2021 11:00',
                },
            ]}
        />,
    )

    expect(wrapper).toMatchSnapshot()
})

it('TemperatureChart no fullscreen button', () => {
    const wrapper = shallow(
        <TemperatureChart2
            allowFullScreen={false}
            sensorData={[
                {
                    d: [
                        18.1,
                        5.3,
                    ],
                    t: '29.03.2021 10:40',
                },
                {
                    d: [
                        18.1,
                        5.3,
                    ],
                    t: '29.03.2021 10:50',
                },
                {
                    d: [
                        18.2,
                        5.3,
                    ],
                    t: '29.03.2021 11:00',
                },
            ]}
        />,
    )

    expect(
        wrapper.find(FullScreen).length,
    ).toEqual(0)
    expect(
        wrapper.find(LineChart).length,
    ).toEqual(1)
})
