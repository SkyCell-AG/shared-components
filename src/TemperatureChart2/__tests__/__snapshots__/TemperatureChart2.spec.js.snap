// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemperatureChart success 1`] = `
<Fragment>
  <div
    className="makeStyles-wrapper-1"
  >
    <FullScreen
      allowFullScreen={true}
      className=""
      customClassName=""
      customColumns={Array []}
      customData={Array []}
      isChartPrinting={false}
      onError={[Function]}
      onFullScreen={[Function]}
      sensorData={
        Array [
          Object {
            "d": Array [
              18.1,
              5.3,
            ],
            "t": "29.03.2021 10:40",
          },
          Object {
            "d": Array [
              18.1,
              5.3,
            ],
            "t": "29.03.2021 10:50",
          },
          Object {
            "d": Array [
              18.2,
              5.3,
            ],
            "t": "29.03.2021 11:00",
          },
        ]
      }
      setTemperatureChartFullscreen={[Function]}
      temperatureChartFullscreen={false}
      userOptions={
        Object {
          "axisTitlesPosition": "none",
          "chartArea": Object {
            "height": "87%",
            "left": 30,
            "top": 20,
            "width": "95%",
          },
          "curveType": "function",
          "series": Array [
            Object {
              "color": "transparent",
              "lineWidth": 0,
              "pointSize": 0,
              "targetAxisIndex": 0,
              "tooltip": false,
              "visibleInLegend": false,
            },
            Object {
              "color": "#CF3B8A",
              "lineWidth": 3,
              "pointSize": 6,
              "targetAxisIndex": 0,
            },
            Object {
              "color": "#61C6E9",
              "lineWidth": 3,
              "pointSize": 6,
              "targetAxisIndex": 0,
            },
            Object {
              "color": "#edae49",
              "lineDashStyle": Array [
                4,
                4,
              ],
              "lineWidth": 3,
              "targetAxisIndex": 0,
            },
            Object {
              "color": "#edae49",
              "lineDashStyle": Array [
                4,
                4,
              ],
              "lineWidth": 3,
              "targetAxisIndex": 0,
            },
          ],
          "vAxis": Object {
            "gridlines": Object {
              "count": 0,
            },
            "title": "Temperature",
          },
        }
      }
    >
      <LineChart
        className="makeStyles-chart-2"
        columns={
          Array [
            Array [
              "datetime",
              "Date",
            ],
            Array [
              "number",
              "Filler",
            ],
          ]
        }
        onError={[Function]}
        userOptions={
          Object {
            "axisTitlesPosition": "none",
            "chartArea": Object {
              "height": "87%",
              "left": 30,
              "top": 20,
              "width": "95%",
            },
            "curveType": "function",
            "series": Array [
              Object {
                "color": "transparent",
                "lineWidth": 0,
                "pointSize": 0,
                "targetAxisIndex": 0,
                "tooltip": false,
                "visibleInLegend": false,
              },
              Object {
                "color": "#CF3B8A",
                "lineWidth": 3,
                "pointSize": 6,
                "targetAxisIndex": 0,
              },
              Object {
                "color": "#61C6E9",
                "lineWidth": 3,
                "pointSize": 6,
                "targetAxisIndex": 0,
              },
              Object {
                "color": "#edae49",
                "lineDashStyle": Array [
                  4,
                  4,
                ],
                "lineWidth": 3,
                "targetAxisIndex": 0,
              },
              Object {
                "color": "#edae49",
                "lineDashStyle": Array [
                  4,
                  4,
                ],
                "lineWidth": 3,
                "targetAxisIndex": 0,
              },
            ],
            "vAxis": Object {
              "gridlines": Object {
                "count": 0,
              },
              "title": "Temperature",
            },
          }
        }
        value={
          Array [
            Array [
              "29.03.2021 10:40",
              0,
              18.1,
              null,
              null,
              5.3,
              null,
              null,
            ],
            Array [
              "29.03.2021 10:50",
              0,
              18.1,
              null,
              null,
              5.3,
              null,
              null,
            ],
            Array [
              "29.03.2021 11:00",
              0,
              18.2,
              null,
              null,
              5.3,
              null,
              null,
            ],
          ]
        }
      />
    </FullScreen>
  </div>
</Fragment>
`;
